### DISCUSSION TITLE
[Help] MDVRP with WorkingTime TimeWindows, but DrivingTime Break

### DISCUSSION THREAD

**Discussion by kd509** [google/or-tools#2972] [Routing (and legacy CP) questions]
Hello,

I started with or-tools some days ago in C# and have a problem now with breaks. Maybe somebody can help me.

I try to solve a routing problem (MDVRP with TimeWindows and Capacity). The TimeWindows are based on working time, which means driving time + service time. So far so good.

What I try to implement now, is a break based on driving time only. So the driver of a vehicle can use the vehicle about 6 hours, but after a maximum of 4.5 hours of driving (!) he needs a break of 45 minutes.

I create the break per vehicle in the following way:
```cs
IntervalVarVector breaks = new IntervalVarVector();
IntervalVar breakInterval = solver.MakeFixedDurationIntervalVar(theBreak.timeWindow.Start, theBreak.timeWindow.End, theBreak.Duration, false, $"Break_{b}_{vehicle.Id}");
RoutingDimension rdDrivingTime = routing.GetMutableDimension("drivingTime");
breaks.Add(breakInterval);
rdDrivingTime.SetBreakIntervalsOfVehicle(breaks, i, node_visit_transits);
```

But how to tell the dimension "workingTime", to insert a break at the correct time, because my break bases on drivingTime?

Regards

Karsten

**Comment by Steffaan** [comment]
Hi Karsten,

Did you solve this problem. I'm also looking for a solutions to the EU driver time regulations in .NET?

Regards,

Steffaan

**Comment by kd509** [comment]
> Hi Karsten,
> 
> Did you solve this problem. I'm also looking for a solutions to the EU driver time regulations in .NET?
> 
> Regards,
> 
> Steffaan

No, I guess it is not possible with or-tools until the bug #1730 is not solved.

Regards

Karsten

**Comment by chris-eaton** [comment]
> Hi Karsten,
> 
> Did you solve this problem. I'm also looking for a solutions to the EU driver time regulations in .NET?
> 
> Regards,
> 
> Steffaan

HI Steffaan, did you ever find a solution? I am struggling to get this working in .NET.

**Comment by DanielaGuerinoni** [comment]
Hi All, any update on this?

**Comment by metaism-ding** [comment]
Any solution for this MDVRPTW ?

**Comment by igorbanfi** [comment]
I have tried by creating a separate dimension used for breaks using driving time, and adding constraint so in each node the slackVar in newly created dimension is <= than slack in time dimension..

```
_node = manager.IndexToNode
model.AddDimensionWithVehicleTransitAndCapacity(
    [model.RegisterTransitCallback(lambda _i, _j:
                                    drive_time[mat, _node(_i), _node(_j)]) for mat in vehicle.mat],
    break_duration,
    manager.GetNumberOfVehicles()
    * [max_drive_time * 2 + break_duration * 3],
    True,
    "VehicleDriveRestTime")

dim_rest_time: RoutingDimension = model.GetDimensionOrDie("VehicleDriveRestTime")
dim_time: RoutingDimension = model.GetDimensionOrDie("Time")
for _i in range(model.Size()):    
    solver.Add(
            dim_rest_time.SlackVar(_i)
            <= dim_time.SlackVar(_i)
        )
```
        
This implementation does need some postprocessing in order to make breaks usable (breaks are set in VehicleDriveRestTime dimension..).

Seems to work, but there is a significant performance drop compared to normal implementation..
